### PR TITLE
feat: support post tags

### PR DESCRIPTION
**Pull request by aoiste**
Support adding tags to posts

Usage:
```md
---
tags: ["deno"]
---
```

or
```md
---
tags:
  - deno
---
```

Preview
image
image

### CODE REVIEW

File: assert.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
+interface Assert<T> {
```
Comment (bartlomieju) [inline assert.ts:1]: This is completely unrelated to this PR, I suggest to revert it

File: blog.tsx [hunk 2]
```diff
@@ -219,7 +227,7 @@ async function loadPost(postsDirectory: string, path: string) {
   pathname = pathname.slice(0, -3);
 
   const { content, data } = frontMatter(contents) as {
-    data: Record<string, string>;
+    data: Record<string, string | string[] | Date>;
```
Comment (bartlomieju) [inline blog.tsx:230]: Why `Date`?

File: components.tsx [hunk 3]
```diff
@@ -263,6 +266,21 @@ function PrettyDate({ date, timezone }: { date: Date; timezone?: string }) {
   return <time dateTime={date.toISOString()}>{formatted}</time>;
 }
 
+function Tags({ tags }: { tags?: string[] }) {
+  return (
+    <section class="flex gap-x-2 flex-wrap">
```
Comment (bartlomieju) [inline components.tsx:271]: Won't this cause to render empty space if no tags are present? I suggest to only render this component if tags are present

File: post.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import type { Post } from "./types.d.ts";
+
+export function filterPosts(
```
Comment (bartlomieju) [inline post.ts:3]: Move contents of this file to `blog.tsx`

**Review by bartlomieju — CHANGES_REQUESTED** [review]

File: assert.ts (hunk 1 above)
Comment (aoiste) [reply assert.ts]: Okay, I will revert this change

File: blog.tsx (hunk 2 above)
Comment (aoiste) [reply blog.tsx]: `frontMatter` actually returns `publish_date` as a  `Date`

File: components.tsx (hunk 3 above)
Comment (aoiste) [reply components.tsx]: Okay, I will fix this

**Review by bartlomieju — COMMENTED** [review]
Looks okay to me, @ije can you take a look too?

File: blog.tsx [hunk 5]
```diff
@@ -464,3 +469,24 @@ export function redirects(redirectMap: Record<string, string>): BlogMiddleware {
...
+  const tag = searchParams.get("tag");
+  if (!tag) {
+    return posts;
+  }
+  return new Map(
+    Array.from(posts.entries()).filter(([, p]) => p.tags?.includes(tag)),
+  );
+}
+
+function recordGetter(data: Record<string, unknown>) {
```
Comment (ije) [inline blog.tsx:486]: seems we don't need this helper

Comment (aoiste) [reply blog.tsx:486]: This helper is used to pass the type check

Comment (aoiste) [reply blog.tsx:486]: The receiver want a `string`, but `data` may give a `string | string[]`

**Review by ije — APPROVED** [review]
LGTM, can you please merge the main? thanks!
